src/lightning/pytorch/cli.py

```diff
@@ -216,50 +217,71 @@ def __init__(
-        fs = get_filesystem(log_dir)
-
-        if not self.overwrite:
-            # check if the file exists on rank 0
-            file_exists = fs.isfile(config_path) if trainer.is_global_zero else False
-            # broadcast whether to fail to all ranks
-            file_exists = trainer.strategy.broadcast(file_exists)
-            if file_exists:
-                raise RuntimeError(
-                    f"{self.__class__.__name__} expected {config_path} to NOT exist. Aborting to avoid overwriting"
-                    " results of a previous run. You can delete the previous config file,"
-                    " set `LightningCLI(save_config_callback=None)` to disable config saving,"
-                    ' or set `LightningCLI(save_config_kwargs={"overwrite": True})` to overwrite the config file.'
+        if self.save_to_log_dir:
+            log_dir = trainer.log_dir  # this broadcasts the directory
+            assert log_dir is not None
+            config_path = os.path.join(log_dir, self.config_filename)
+            fs = get_filesystem(log_dir)
+
+            if not self.overwrite:
+                # check if the file exists on rank 0
+                file_exists = fs.isfile(config_path) if trainer.is_global_zero else False
+                # broadcast whether to fail to all ranks
+                file_exists = trainer.strategy.broadcast(file_exists)
+                if file_exists:
+                    raise RuntimeError(
+                        f"{self.__class__.__name__} expected {config_path} to NOT exist. Aborting to avoid overwriting"
+                        " results of a previous run. You can delete the previous config file,"
+                        " set `LightningCLI(save_config_callback=None)` to disable config saving,"
+                        ' or set `LightningCLI(save_config_kwargs={"overwrite": True})` to overwrite the config file.'
+                    )
+
+            if trainer.is_global_zero:
+                # save only on rank zero to avoid race conditions.
+                # the `log_dir` needs to be created as we rely on the logger to do it usually
+                # but it hasn't logged anything at this point
+                fs.makedirs(log_dir, exist_ok=True)
+                self.parser.save(
+                    self.config, config_path, skip_none=False, overwrite=self.overwrite, multifile=self.multifile
                 )
 
-        # save the file on rank 0
         if trainer.is_global_zero:
-            # save only on rank zero to avoid race conditions.
-            # the `log_dir` needs to be created as we rely on the logger to do it usually
-            # but it hasn't logged anything at this point
-            fs.makedirs(log_dir, exist_ok=True)
-            self.parser.save(
-                self.config, config_path, skip_none=False, overwrite=self.overwrite, multifile=self.multifile
-            )
+            self.save_config(trainer, pl_module, stage)
```

Comment:
Since `save_config` is called only on global zero, any collective call inside will make it hang.
A simple example being that a users calls `trainer.log_dir`.

This would be very limiting when `save_to_log_dir is False`. Why don't we let it run on all ranks?

Reply:
> Is there any way (e.g. context manager) to make it such that if someone tries to make a collective call, it simply fails instead of hanging?

Not at the moment.

Resolving